consensus/src/alerts/handler.rs

```diff
@@ -0,0 +1,786 @@
+>;
+
+type OnMessageResult<H, D, MK> = Result<
+    Option<
+        AlerterResponse<
+            H,
+            D,
+            <MK as Keychain>::Signature,
+            <MK as MultiKeychain>::PartialMultisignature,
+        >,
+    >,
+    Error,
+>;
+
+/// The component responsible for fork alerts in AlephBFT. We refer to the documentation
+/// https://cardinal-cryptography.github.io/AlephBFT/how_alephbft_does_it.html Section 2.5 and
+/// https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html and to the Aleph
+/// paper https://arxiv.org/abs/1908.05156 Appendix A1 for a discussion.
+pub struct Handler<H: Hasher, D: Data, MK: MultiKeychain> {
+    session_id: SessionId,
+    keychain: MK,
+    known_forkers: HashMap<NodeIndex, ForkProof<H, D, MK::Signature>>,
+    known_alerts: KnownAlerts<H, D, MK>,
+    known_rmcs: HashMap<(NodeIndex, NodeIndex), H::Hash>,
+    pub exiting: bool,
+}
+
+impl<H: Hasher, D: Data, MK: MultiKeychain> Handler<H, D, MK> {
+    pub fn new(keychain: MK, config: AlertConfig) -> Self {
+        Self {
+            session_id: config.session_id,
+            keychain,
+            known_forkers: HashMap::new(),
+            known_alerts: HashMap::new(),
+            known_rmcs: HashMap::new(),
+            exiting: false,
+        }
+    }
+
+    pub fn index(&self) -> NodeIndex {
+        self.keychain.index()
+    }
+
+    pub fn is_forker(&self, forker: NodeIndex) -> bool {
+        self.known_forkers.contains_key(&forker)
+    }
+
+    pub fn on_new_forker_detected(
+        &mut self,
+        forker: NodeIndex,
+        proof: ForkProof<H, D, MK::Signature>,
+    ) {
+        self.known_forkers.insert(forker, proof);
+    }
+
+    // Correctness rules:
+    // 1) All units must be created by forker
+    // 2) All units must come from different rounds
+    // 3) There must be fewer of them than the maximum defined in the configuration.
+    // Note that these units will have to be validated before being used in the consensus.
+    // This is alright, if someone uses their alert to commit to incorrect units it's their own
+    // problem.
+    pub fn verify_commitment(&self, alert: &Alert<H, D, MK::Signature>) -> Result<(), Error> {
+        let mut rounds = HashSet::new();
+        for u in &alert.legit_units {
+            let u = match u.clone().check(&self.keychain) {
+                Ok(u) => u,
+                Err(_) => return Err(Error::IncorrectlySignedUnit(alert.sender)),
+            };
+            let full_unit = u.as_signable();
+            if full_unit.creator() != alert.forker() {
+                return Err(Error::WrongCreator(alert.sender));
+            }
+            if rounds.contains(&full_unit.round()) {
+                return Err(Error::SameRound(full_unit.round(), alert.sender));
+            }
+            rounds.insert(full_unit.round());
+        }
+        Ok(())
+    }
+
+    fn verify_fork(&self, alert: &Alert<H, D, MK::Signature>) -> Result<NodeIndex, Error> {
```

Comment:
You're probably right. I'd say the same about `verify_commitment`. But these methods can return `handler::Error` and it would be weird to me if `Alert` depended on this error type. We could move `handler::Error` along with these methods to `alerts/mod` but this doesn't seem ideal to me as I see it fit much better inside `handler`.

Reply:
Maybe we just leave it as is in this PR.